lib/internal/webstreams/adapters.js

```diff
@@ -0,0 +1,883 @@
+  const strategy =
+    objectMode ?
+      new CountQueuingStrategy({ highWaterMark }) :
+      { highWaterMark };
+
+  let controller;
+  let closed = false;
+
+  function maybeCloseOrError(error) {
+    if (closed) return;
+    closed = true;
+    if (error)
+      return controller.error(error);
+    controller.close();
+  }
+
+  function onClose() {
+    streamReadable.off('close', onClose);
+    streamReadable.off('data', onData);
+    streamReadable.off('end', onEnd);
+    streamReadable.off('error', onError);
+    maybeCloseOrError();
+  }
+
+  function onError(error) {
+    // If closed is false, the stream.Readable errored
+    // before the ReadableStream was closed, we need to
+    // go ahead and error it out.
+    maybeCloseOrError(error);
+  }
+
+  function onEnd() {
+    // If closed is false, the stream.Readable is ending
+    // before the ReadableStream, we need to go ahead an
+    // close the ReadableStream.
+    maybeCloseOrError();
+  }
+
+  function onData(chunk) {
+    // Copy the Buffer to detach it from the pool.
+    if (Buffer.isBuffer(chunk) && !objectMode)
+      chunk = new Uint8Array(chunk);
+    controller.enqueue(chunk);
+  }
+
+  streamReadable.pause();
+
+  streamReadable.once('close', onClose);
+  streamReadable.once('end', onEnd);
+  streamReadable.once('error', onError);
+  streamReadable.on('data', onData);
+
+  return new ReadableStream({
+    start(c) { controller = c; },
+
+    pull() { streamReadable.resume(); },
+
+    cancel(reason) {
+      closed = true;
+      // If the streamReadable has not yet been destroyed,
+      // destroy it immediately with the given reason.
+      if (!streamReadable.destroyed)
+        streamReadable.destroy(reason);
+    },
+  }, strategy);
+}
+
+/**
+ * @param {ReadableStream} readableStream
+ * @param {{
+ *   highWaterMark? : number,
+ *   encoding? : string,
+ *   objectMode? : boolean,
+ *   signal? : AbortSignal,
+ * }} [options]
+ * @returns {Readable}
+ */
+function newStreamReadableFromReadableStream(readableStream, options = {}) {
+  if (!isReadableStream(readableStream)) {
+    throw new ERR_INVALID_ARG_TYPE(
+      'readableStream',
+      'ReadableStream',
+      readableStream);
+  }
+
+  validateObject(options, 'options');
+  const {
+    highWaterMark,
+    encoding,
+    objectMode = false,
+    signal,
+  } = options;
+
+  if (encoding !== undefined && !Buffer.isEncoding(encoding))
+    throw new ERR_INVALID_ARG_VALUE(encoding, 'options.encoding');
+  validateBoolean(objectMode, 'options.objectMode');
+  if (signal !== undefined)
+    validateAbortSignal(signal, 'options.signal');
+
+  const reader = readableStream.getReader();
+  let closed = false;
+
+  const readable = new Readable({
+    objectMode,
+    highWaterMark,
+    encoding,
+    signal,
+
+    read() {
+      PromisePrototypeThen(
+        reader.read(),
+        (chunk) => {
+          readable.push(chunk.value);
+          if (chunk.done)
+            readable.push(null);
```

Comment:
If `chunk.done === true`, I believe this is pushing *both* `undefined` and `null`. While pushing `undefined` is allowed, it's not recommended (see docs on [readable.push()](https://nodejs.org/api/stream.html#stream_readable_push_chunk_encoding) and [readable.push('')](https://nodejs.org/api/stream.html#stream_readable_push)).

We should only push `null`.
```suggestion
          if (chunk.done) {
            readable.push(null);
          } else {
            readable.push(chunk.value);
          }
```